### ISSUE TITLE
There is no way to access request's raw body

### ISSUE THREAD

**Issue by antonsapyanov** [actionhero/actionhero#1081]
I implement an integration with Stripe, and need access to request's raw body in order to verify signature (Stripe makes POST requests on specified webhook).
I have found [that](https://github.com/actionhero/actionhero/blob/v16.0.0/servers/web.js#L484) in the source code:
```
connection.rawConnection.form = new formidable.IncomingForm()

for (i in api.config.servers.web.formOptions) {
  connection.rawConnection.form[i] = api.config.servers.web.formOptions[i]
}

connection.rawConnection.form.parse(connection.rawConnection.req, (error, fields, files) => {
  if (error) {
    server.log('error processing form: ' + String(error), 'error')
    connection.error = new Error('There was an error processing this form.')
  } else {
    connection.rawConnection.params.body = fields
    connection.rawConnection.params.files = files
    fillParamsFromWebRequest(connection, files)
    fillParamsFromWebRequest(connection, fields)
  }
  if (api.config.servers.web.queryRouting !== true) { connection.params.action = null }
  api.routes.processRoute(connection, pathParts)
  callback(requestMode)
})
```
As I understood, there is no way to access raw body via `rawConnection` object. Is there any other way (maybe suggestions) how to access it?

- ActionHero Version: `16.0.0`

**Comment by evantahler** [comment]
If you are talking about a JSON body payload, ie: `curl -H "Content-Type: application/json" -X POST -d '{"username":"xyz","password":"xyz"}' http://localhost:3000/api/login`, you can use `connection.rawConnection.params.body`.

**Comment by evantahler** [comment]
keep in mind that `data.connection.rawConnection.params.body` differs from `data.params` in a few ways:
- no formatters, validators, defaults, etc will be applied
- no sanitization will be done on them, so be sure that you are testing that the request is coming from a trusted source (ie: only stripe's servers)

**Comment by antonsapyanov** [comment]
I mean exactly **raw body** (String or Buffer), before it is parsed to an object.

**Comment by antonsapyanov** [comment]
Stripe sends JSON object. There is no reason to stringify `data.connection.rawConnection.params.body`, because there is no guarantee, that result string will equal request's raw body string (which is used to make Stripe Signature header).

**Comment by antonsapyanov** [comment]
Could you reopen the issue please? I think it is quite important

**Comment by evantahler** [comment]
Can you provide as test CURL request with a body payload which emulates the response you are trying to access from ActionHero?

**Comment by evantahler** [comment]
realted to https://github.com/actionhero/actionhero/issues/213 and https://github.com/actionhero/actionhero/pull/318

**Comment by evantahler** [comment]
@antonsapyanov the curl statement will be very helpful.  For example, this is invalid `curl -X POST -H "Content-Type: text/plain" --data "this is raw data" http://localhost:8080/api/randomnumber` because you can't have POST data which is just a string.  You can have POST data with other content-types, but those are already parsed, as shown above

**Comment by evantahler** [comment]
Also, `https://stripe.com/docs/webhooks#signatures` seems to indicate the data you need for verification is actually in a request header, which you can get via `onnection.rawConnection.req.headers`

**Comment by antonsapyanov** [comment]
Here is a simulation of request, that Stripe sends:
```
curl -X POST -H "Content-Type: application/json" -H "Stripe-Signature: <STRIPE_SIGNATURE_MADE_FROM_REQUEST_BODY_CONTENT>" -d '{"id":"evt_1AdxX6JJkD9VAcbRjTfTLfBi","data":{"object":{"id":"ii_1AdxX6JJkD9VAcbR6VmutlEc","object":"invoiceitem","amount":99972,"currency":"usd","customer":"cus_AzxFyutLWWVW0x","date":1499676380,"description":"Remaining time on Enterprise after 10 Jul 2017","discountable":false,"invoice":null,"livemode":false,"metadata":{},"period":{"start":1499676380,"end":1502354032},"plan":{"id":"3","object":"plan","amount":100000,"created":1497342793,"currency":"usd","interval":"month","interval_count":1,"livemode":false,"metadata":{},"name":"Enterprise","statement_descriptor":null,"trial_period_days":null},"proration":true,"quantity":1,"subscription":"sub_AzxFTHB8YZx78C","subscription_item":"si_1AdxL2JJkD9VAcbRf8uStNkt"}},"type":"invoiceitem.created"}' http://localhost:8080/api/stripe_handler
```
To verify the request, I need to call:
```
stripe.webhooks.constructEvent(<RAW_REQUEST_BODY>, <STRIPE_SIGNATURE_HEADER>, <SIGNING_SECRET>)
```
There is exactly what I call:
```
var sign = data.connection.rawConnection.req.headers['stripe-signature'];
try {
  var event = api.stripe.webhooks.constructEvent(JSON.stringify(data.connection.rawConnection.params.body), sign, api.config.stripe.sign_secret);
} catch (e) {
  next(e);
}
```
And that results with an exception: `'No signatures found matching the expected signature for payload'`.

**Comment by antonsapyanov** [comment]
To successfully pass the try-catch block, I need to provide not `JSON.stringify(data.connection.rawConnection.params.body)`, but exactly the request`s raw body, which is the JSON object (as a String) sent by Stripe.

**Comment by evantahler** [comment]
Got it!  Check out https://github.com/actionhero/actionhero/pull/1082... I think this would solve your problem?

**Comment by antonsapyanov** [comment]
Thanks! It is exactly what we need. Waiting for successful pull request.

**Comment by evantahler** [comment]
closed by #1082
